### PR TITLE
 feat(socketioxide/v4): implement SocketIO V4 (closes #51)

### LINKED ISSUE #51: Add SocketIO v4 support
Created this issue as a way to keep track of this feature.

Currently only v5 - which uses EngineIO v4 under the hood - is implemented. Adding support for v4 - which is based on EngineIO v3 - should be possible now that #22 has been resolved.

Differences between [v5 and v4](https://github.com/socketio/socket.io-protocol/tree/main#difference-between-v5-and-v4) are also fairly small in scope.

### PR DESCRIPTION
**Pull request by sleeyax**
Protocol implementation steps:

- [x] Add implicit connection to the default namespace
- ~~rename `CONNECT_ERROR` to `ERROR`~~ Not necessary because the value hasn't changed, only the name. 
- [x] The `CONNECT` packet shouldn't contain a payload
- [ ] The payload `CONNECT_ERROR` should be a plain string instead of an object. *I might skip implementing this, I think a v4 client currently would receive a JSON string as the error, which isn't a big deal.*
- [x] Add E2E tests (WIP: https://github.com/sleeyax/socket.io-protocol/tree/v4-testsuite)

### CODE REVIEW

**Comment by sleeyax** [comment]
I'm having a hard time figuring out an annoying issue with [tests](https://github.com/sleeyax/socket.io-protocol/tree/v4-testsuite). Sometimes the tests fail and sometimes they don't :thinking: 

It's likely related to this part of the implementation:

>  Add implicit connection to the default namespace

I'm not sure how tests should behave when the client is already connected to the default namespace when the connection is made. Maybe I'm encountering a race condition? Or perhaps my server implementation is just wrong. Either way, I would appreciate it if someone with some time to spare could have a look and help me out, I need to take a break from this :)

**Example to reproduce**

Server:

`cargo run --bin socketioxide-e2e --all-features `

Client:

`npm test -- -f "should allow connection to a custom namespace"`

```js
    // ...
    it("should allow connection to a custom namespace", async () => {
      for (let i = 0; i < 3; i++) {
        console.log(i); // works at i = 0, fails at i = 1

        const socket = new WebSocket(
          `${WS_URL}/socket.io/?EIO=3&transport=websocket`
        );
  
        await waitFor(socket, "message"); // Engine.IO handshake

        socket.send("40/custom");
  
        const { data } = await waitFor(socket, "message");
  
        expect(data).to.startsWith("40/custom");
      }
    });
```

File: socketioxide/src/client.rs [hunk 1]
```diff
@@ -59,16 +59,17 @@ impl<A: Adapter> Client<A> {
...
         socket: &EIoSocket<Self>,
     ) -> Result<(), SendError> {
         debug!("auth: {:?}", auth);
         let handshake = Handshake::new(auth, socket.req_data.clone());
         let sid = socket.sid;
         if let Some(ns) = self.get_ns(&ns_path) {
+            let protocol = socket.protocol;
             let socket = ns.connect(sid, socket.tx.clone(), handshake, self.config.clone());
-            socket.send(Packet::connect(ns_path.clone(), sid))?;
+            socket.send(Packet::connect(ns_path.clone(), sid, protocol))?;
```
Comment (Totodore) [inline socketioxide/src/client.rs:72]: You need to send the connect response only when it is not a _default ns connection_. I think this is the cause of some of your issues
```suggestion
  if protocol == engineioxide::service::ProtocolVersion::V4 || ns_path != "/" {
      socket.send(Packet::connect(ns_path.clone(), sid, protocol))?;
  }
```

**Comment by sleeyax** [comment]
Hey all, just wanted to mention that I'm currently busy with other projects. I'm not sure when I'll be able to get back to this PR. If anyone wants to take over, definitely don't wait for me. Cheers!

**Comment by Totodore** [comment]
> I'm having a hard time figuring out an annoying issue with [tests](https://github.com/sleeyax/socket.io-protocol/tree/v4-testsuite). Sometimes the tests fail and sometimes they don't 🤔
> 
> It's likely related to this part of the implementation:
> 
> > Add implicit connection to the default namespace
> 
> I'm not sure how tests should behave when the client is already connected to the default namespace when the connection is made. Maybe I'm encountering a race condition? Or perhaps my server implementation is just wrong. Either way, I would appreciate it if someone with some time to spare could have a look and help me out, I need to take a break from this :)
> 
> **Example to reproduce**
> 
> Server:
> 
> `cargo run --bin socketioxide-e2e --all-features `
> 
> Client:
> 
> `npm test -- -f "should allow connection to a custom namespace"`
> 
> ```js
>     // ...
>     it("should allow connection to a custom namespace", async () => {
>       for (let i = 0; i < 3; i++) {
>         console.log(i); // works at i = 0, fails at i = 1
> 
>         const socket = new WebSocket(
>           `${WS_URL}/socket.io/?EIO=3&transport=websocket`
>         );
>   
>         await waitFor(socket, "message"); // Engine.IO handshake
> 
>         socket.send("40/custom");
>   
>         const { data } = await waitFor(socket, "message");
>   
>         expect(data).to.startsWith("40/custom");
>       }
>     });
> ```

A bit late but it correspond to this issue : https://github.com/socketio/socket.io-protocol/issues/32

**Comment by tausifcreates** [comment]
Is it necessary to support a legacy protocol?

**Comment by Totodore** [comment]
> Is it necessary to support a legacy protocol?

@sleeyax made 90% of the work, it is not a big deal to merge this PR

**Comment by sleeyax** [comment]
> Is it necessary to support a legacy protocol?

I personally needed this to support a legacy socketio client that pretty much can't upgrade to the latest protocol version any time soon. I'm sure others will have similar use cases in the future for socketioxide and given you can leverage Rust feature flags I don't see a reason not to support older protocols. And finally, like totodore said most of the work for V4 is done already anyways :)

**Comment by tausifcreates** [comment]
> > Is it necessary to support a legacy protocol?
> 
> I personally needed this to support a legacy socketio client that pretty much can't upgrade

Makes perfect sense, thanks for explaining @sleeyax

**Pull request merged**